### PR TITLE
[Fix](bloom filter) Fix bloom filter memory leak

### PR DESCRIPTION
**Pull request by Yukang-Lian**
## Proposed changes

Issue Number: close #xxx

Issue: Doris occasionally encounters an issue where memory usage becomes exceptionally high and does not decrease. The leaked memory is occupied by Bloom filters stored in memory.

Reason: The segment cache stores segment objects read from files into memory. It functions as an LRU cache with an eviction strategy: when the number of segments exceeds the maximum number, or the total memory size of segment objects in the cache exceeds the maximum usage, it evicts the older segments. However, there is a piece of logic in the code that first reads the segment object into memory, assuming it occupies memory size A, then places the read segment object into the cache (at this point, the cache considers the segment object size to be A). It then reads the segment's Bloom filter from the file and assigns it to the segment's Bloom filter member variable, assuming the Bloom filter occupies memory size B. Thus, the total size of the segment object at this point is A+B. However, the cache does not update this size, leading to the actual size of the segment object stored in the cache (A+B) being larger than the size considered by the cache (A). When the number of segment objects in the cache increases to a certain extent, the used memory will surge dramatically. However, the cache does not perceive the size as reaching the eviction limit, so it does not evict the segment objects. In such cases, a memory leak issue arises.

Solution: Since each segment object only reads the Bloom filter once, the issue can be resolved by changing the logic from reading the segment, placing it into the cache, and then reading the Bloom filter to reading the segment, reading the Bloom filter, and then placing it into the cache.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

File: be/src/olap/rowset/segment_v2/segment.cpp [hunk 1]
```diff
@@ -290,6 +290,10 @@ Status Segment::_load_pk_bloom_filter() {
         return _load_pk_bf_once.call([this] {
             RETURN_IF_ERROR(_pk_index_reader->parse_bf(_file_reader, *_pk_index_meta));
             _meta_mem_usage += _pk_index_reader->get_bf_memory_size();
+            SegmentCache::CacheKey cache_key(_rowset_id, _segment_id);
```
Comment (zhannngchen) [inline be/src/olap/rowset/segment_v2/segment.cpp:293]: It's better to decouple cache management and load index/bloomfilter

Comment (zhannngchen) [reply be/src/olap/rowset/segment_v2/segment.cpp:293]: Sometimes, we might call `_load_pk_bloom_filter()` on `Segment` objects that not in cache. Not all Segment object should in SegmentCache

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

File: be/src/olap/base_tablet.cpp [hunk 2]
```diff
@@ -548,6 +553,15 @@ Status BaseTablet::lookup_row_key(const Slice& encoded_key, bool with_seq_col,
 
         for (auto id : picked_segments) {
             Status s = segments[id]->lookup_row_key(encoded_key, with_seq_col, with_rowid, &loc);
+            SegmentCache::CacheKey cache_key(rs->rowset_id(), segments[id]->id());
+            auto cache_value = std::make_unique<SegmentCache::CacheValue>();
+            cache_value->segment = segments[id];
+            try {
+                SegmentLoader::instance()->update_segment(cache_key, *cache_value);
```
Comment (zhannngchen) [inline be/src/olap/base_tablet.cpp:560]: `load_pk_index_and_bf` only would be called once, but `lookup_row_key` might be called for billion times, if you also call `update_segment` for so many times ,the LRU write lock would become bottleneck.

**Comment by Yukang-Lian** [comment]
run buildall

File: be/src/olap/base_tablet.cpp [hunk 3]
```diff
@@ -538,10 +538,10 @@ Status BaseTablet::lookup_row_key(const Slice& encoded_key, bool with_seq_col,
             continue;
         }
 
-        if (UNLIKELY(segment_caches[i] == nullptr)) {
+        if (LIKELY(segment_caches[i] == nullptr)) {
```
Comment (zhannngchen) [inline be/src/olap/base_tablet.cpp:541]: should be UNLIKELY

**Review by zhannngchen — DISMISSED** [review]
LGTM

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run p0

**Review by zhannngchen — DISMISSED** [review]

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Review by zhannngchen — APPROVED** [review]
LGTM

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**